iml-services/iml-stats/src/main.rs

```diff
@@ -531,51 +532,97 @@ async fn main() -> Result<(), ImlStatsError> {
     Ok(())
 }
 
-#[derive(serde::Serialize, serde::Deserialize)]
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
 struct MgsFsTime {
     time: u64,
+    mgs_fs: String,
+    mgs_fs_count: Option<u32>,
 }
 
 async fn delete_existing_mgs_fs_records(
-    fs_names: String,
+    fs_names: HashSet<String>,
     client: &Client,
 ) -> Result<(), ImlStatsError> {
     let xs: Vec<MgsFsTime> = client
         .query_into(
             format!(
                 r#"
-            SELECT mgs_fs,is_mgs_fs
-            FROM target
-            WHERE mgs_fs='{}' AND is_mgs_fs=true
-            ORDER BY time ASC
-            LIMIT 2"#,
-                fs_names
+                SELECT mgs_fs, mgs_fs_count FROM
+                    (SELECT count(is_mgs_fs) AS mgs_fs_count FROM target WHERE is_mgs_fs=true GROUP BY mgs_fs),
+                    (SELECT LAST(is_mgs_fs),time as last_time FROM target GROUP BY mgs_fs)
+                GROUP BY mgs_fs
+                "#
             )
             .as_str(),
-            Some(Precision::Nanoseconds),
+            Some(Precision::Nanoseconds)
         )
         .await?
         .unwrap_or_default();
 
-    if xs.len() < 2 {
-        return Ok(());
-    }
+    tracing::debug!(
+        "delete stale mgs_fs records - fs_names: {:?}; query results: {:?}",
+        fs_names,
+        xs
+    );
 
-    if let Some(x) = xs.get(0) {
-        let query = format!(
-            "DELETE FROM target WHERE time < {} AND mgs_fs = '{}' AND kind = '{}'",
-            x.time,
-            fs_names,
-            TargetVariant::MGT
-        );
+    let subsets: HashMap<String, (u64, u32)> = xs
+        .into_iter()
+        .filter(|x| {
+            let mgs_fs_set = fs_names_to_hash_set(&x.mgs_fs);
 
-        tracing::debug!("Running query: {}", query);
+            let intersection = mgs_fs_set.intersection(&fs_names).collect::<Vec<_>>();
 
-        client
-            .query(query.as_str(), Some(Precision::Nanoseconds))
-            .map_err(|e| iml_influx::Error::InfluxDbError(e))
-            .await?;
-    }
+            !intersection.is_empty()
+        })
+        .fold(HashMap::new(), |mut acc, record| {
+            let mgs_fs = record.mgs_fs;
+            let mut entry = acc.entry(mgs_fs).or_insert((0, 0));
+
+            if record.time > 0 {
+                entry.0 = record.time;
+            }
+
+            if let Some(count) = record.mgs_fs_count {
+                entry.1 = count;
+            }
+
+            acc
+        });
+
+    let delete_subset_futures = subsets
+        .into_iter()
+        .filter(|(mgs_fs, (_time, count))| {
+            let mgs_fs_set = fs_names_to_hash_set(&mgs_fs);
+
+            mgs_fs_set != fs_names || *count > 1
```

Comment:
Why is this an or?

Reply:
The filter will continue for anything expression that is true and since I want to filter out anything the following:
```
mgs_fs_set == fs_names && count <= 1
```
The inverse of this expression will accept any case that does not match the above. Here is sample data that is run against expression `mgs_fs_set != fs_names || *count > 1`:
```
mgs_fs_set: {"fs"} fs_names: {"fs2", "fs"}, count: 1, passes filter: true
mgs_fs_set: {"fs"} fs_names: {"fs2", "fs"}, count: 2, passes filter: true
mgs_fs_set: {"fs2"} fs_names: {"fs2", "fs"}, count: 1, passes filter: true
mgs_fs_set: {"fs2"} fs_names: {"fs2", "fs"}, count: 2, passes filter: true
mgs_fs_set: {"fs2", "fs"} fs_names: {"fs2", "fs"}, count: 1, passes filter: false
mgs_fs_set: {"fs2", "fs"} fs_names: {"fs2", "fs"}, count: 2, passes filter: true
mgs_fs_set: {"fs3", "fs", "fs2"} fs_names: {"fs2", "fs"}, count: 1, passes filter: true
mgs_fs_set: {"fs3", "fs", "fs2"} fs_names: {"fs2", "fs"}, count: 2, passes filter: true
```